File: fs/before_checksum/939__valid-permutations-for-di-sequence/cpp/valid-permutations-for-di-sequence_benchmark_v2.cpp
#include <iostream>
#include <vector>
#include <string>

class Solution {
public:
    int numPermsDISequence(std::string S) {
        const int MOD = 1000000007;
        int n = S.size();
        std::vector<int> dp(n + 1, 1);
        for (char c : S) {
            if (c == 'I') {
                dp.pop_back();
                for (int i = 1; i < static_cast<int>(dp.size()); ++i) {
                    dp[i] += dp[i - 1];
                    if (dp[i] >= MOD) dp[i] -= MOD;
                }
            } else {
                std::vector<int> newdp(dp.begin() + 1, dp.end());
                dp.swap(newdp);
                for (int i = static_cast<int>(dp.size()) - 2; i >= 0; --i) {
                    dp[i] += dp[i + 1];
                    if (dp[i] >= MOD) dp[i] -= MOD;
                }
            }
        }
        return dp[0] % MOD;
    }
};

volatile int sink = 0;

int main() {
    std::vector<std::string> tests = {
        "I",
        "D",
        "ID",
        "DI",
        "IID",
        "DDI",
        "IDIDID",
        "DDDDIIII",
        "IIDDDIII",
        "IDDIIDIDID"
    };

    Solution sol;

    const int iterations = 1;
    for (int iter = 0; iter < iterations; ++iter) {
        int checksum = 0;
        for (const auto& s : tests) {
            checksum += sol.numPermsDISequence(s);
        }
        sink += checksum;
        

    } std::cout << sink << std::endl;

    return 0;
}

File: fs/py_before_energy/939__valid-permutations-for-di-sequence/valid-permutations-for-di-sequence_benchmark_v2.py
# Time:  O(n^2)
# Space: O(n)

class Solution(object):
    def numPermsDISequence(self, S):
        """
        :type S: str
        :rtype: int
        """
        dp = [1]*(len(S)+1)
        for c in S:
            if c == "I":
                dp = dp[:-1]
                for i in xrange(1, len(dp)):
                    dp[i] += dp[i-1]
            else:
                dp = dp[1:]
                for i in reversed(xrange(len(dp)-1)):
                    dp[i] += dp[i+1]
        return dp[0] % (10**9+7)

if __name__ == "__main__":
    tests = [
        "I",
        "D",
        "ID",
        "DI",
        "IID",
        "DDI",
        "IDIDID",
        "DDDDIIII",
        "IIDDDIII",
        "IDDIIDIDID"
    ]

    sol = Solution()

    iterations = 1000
    for iter in xrange(iterations):
        #checksum = 0
        for s in tests:
            r = sol.numPermsDISequence(s)
        #sink = checksum